### PR TITLE
Changing the representation of stack allocated arrays to allow for array valued command line arguments

### PR DESCRIPTION
**Pull request by bettinaheim**
This PR changes the representation of stack allocated arrays. The new representation allows to retroactively define the correct array length as part of a subsequent compilation step. For more details, read the comments in the code. 
Existing tests should cover existing functionality. Adding tests for the new functionality requires merging https://github.com/microsoft/qsharp-compiler/pull/1554; I'll add the tests once that PR has merged.

### CODE REVIEW

File: src/QsCompiler/QirGeneration/QIR/Types.cs [hunk 1]
```diff
@@ -142,6 +159,28 @@ internal Types(Context context, Func<QsQualifiedName, QsCustomType?> getTypeDecl
...
+
+            return type;
+        }
+
+        /// <summary>
+        /// Creates the LLVM data type used for stack allocated arrays whose size will be resolved
+        /// as part of a compilation step and only after the initial QIR generation.
+        /// </summary>
+        internal IStructType NamedLlvmArray(string storageName, ITypeRef elementType) =>
+            Emission.ArrayValue.StackAllocatedArrayType(this, elementType, 0u, storageName);
```
Comment (kuzminrobin) [inline src/QsCompiler/QirGeneration/QIR/Types.cs:182]: (nit) Passing a hard-coded value (magic number `0u`) is not very informative. For my code in such cases I prefer to add the param name:
```suggestion
            Emission.ArrayValue.StackAllocatedArrayType(this, elementType, nrElements: 0u, storageName);
```

File: src/QsCompiler/QirGeneration/Subtransformations/TypeTransformation.cs [hunk 2]
```diff
@@ -65,7 +68,9 @@ internal ITypeRef LlvmTypeFromQsharpType(ResolvedType resolvedType)
 
             public override QsResolvedTypeKind OnArrayType(ResolvedType b)
             {
-                this.BuiltType = this.QirTypes.Array;
+                this.BuiltType = this.namedLlvmArrays is null
+                    ? this.QirTypes.Array
+                    : this.QirTypes.NamedLlvmArray($"ArrayStorage{++this.namedLlvmArrays}", this.LlvmTypeFromQsharpType(b));
```
Comment (kuzminrobin) [inline src/QsCompiler/QirGeneration/Subtransformations/TypeTransformation.cs:73]: (nit)
For readability I would recommend (if that compiles):
```suggestion
                    : this.QirTypes.NamedLlvmArray($"ArrayStorage{++(this.namedLlvmArrays)}", this.LlvmTypeFromQsharpType(b));
```

**Review by kuzminrobin — APPROVED** [review]
Approving to unblock, but I would recommend an approval from those familiar with the code base.

**Comment by swernli** [comment]
Closing in favor of #1591.

**Pull request closed without merging**